submariner-operator/diagnose-refactor-ocm.md

```diff
@@ -0,0 +1,151 @@
+# Refactor subctl diagonose for use with OCM
+
+## Summary
+
+Currently only way to run diagnostics on a Submariner installation is through `subctl` CLI. This requires user or admin
+to run it manually from a host. There is no way to run it through UI like OCM or deployment tools like helm charts etc.
+
+## Proposal
+
+This enhancement proposes a means to run `subctl diagnose` through a `SubmarinerDiagnose` Job. User or Admin can deploy
+this Job manually. In case of OCM, a new SubmarinerDiagnose Controller will be added to SubmarinerAddon which will deploy
+this Job, track its completion and fetch results once done.
+
+## Design Details
+
+* A new option `--output json` will be added to `subctl diagnose` to capture output in a json format instead of current
+ user readable text format.
+* `subctl diagnose` will also update some Prometheus metrics whenever it is run.
+* A `submariner-diagnose` Job will be created which will run `subctl diagnose` with relevant flags as `command`. This
+ job will run once to completion. This Job will always run with new option of `--output json`.
+* A new `SubmarinerDiagnose` CRD will be added to SubmarinerAddon. OCM UI will create this CR with relevant values
+ to trigger running diagnose from OCM UI.`
+* A new `SubmarnerDiagnoseController` Controller will be added to OCM SubmarinerAddon. This controller will spawn the SubmarinerDiagnose
+ Job when SubmarinerDiagnose CR is created, track completion.
+* Only one active `submariner-diagnose` Job will be run in a cluster by SubmarinerAddon at a given time. Users can create
+ multiple SubmarinerDiagnose Jobs manually and we won't enforce any checks to prevent that.  
+* When `submariner-diagnose` Job is completed, SubmarinerDiagnoseController will fetch the logs from diagnose Pod, extract the
+ results in json format and populate the relevant Status fields in SubmarinerDiagnose CR.
+* Support will be added to OCM UI to run diagnostics on a cluster with a single click and then display the results. For now,
+ it will run with sub-option `all`
+* Results of diagnose will also be exposed as Prometheus metrics. This will allow any Prometheus based observability
+  solutions to consume the results of diagnose. This will also provide a quick and easy way to show results in the UI while
+  SubmarinerAddon changes are in-progress.
+* [Initial scope] For initial implementation, SubmarinerDiagnose will always run with `all` but without Firewall Options.
+ This is to reduce complexity on OCM UI side to pass credentials for remote cluster which are required for Firewall option.
+ This support will be added in later release.
+
+### Alternate Design
+
+* Instead of `--output json` option to provide results in json format for SubmarinerAddon to consume, it will just retain
+ existing text output.
+* UI will provide a link to logs of `submariner-diagnose` Pod and show results as is to the user.
```

Comment:
This is a bit narrow scope right now, I think if we have this route perhaps it would be nice to let users execute any `subctl` command arbitrarily from the UI, or is it a crazy idea? :)

Reply:
That indeed is where we want to end up with. Option to run any command for "power users" alongwith easy to run options for rest. We're going with this narrow scope for now primarily coz it is easier/faster to do in UI.